### PR TITLE
added formatting procedures pipeline

### PR DESCRIPTION
**Pull request by MkYacine**

### CODE REVIEW

File: dataset/prompt.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
...
+  "task": "Now here is the procedure I want you to format:\n--BEGIN PROCEDURE--\n{}\n--END PROCEDURE--"
+}
```
Comment (kylrth) [inline dataset/prompt.json]: Format this json yo

File: dataset/format_procedure.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+task = prompt["task"]
+TOKEN__LIMIT = 8192
+
+
+def format_procedure(full_procedure: str) -> str:
+    """Formats a single procedure with GPT-4"""
+    temp = task.format(full_procedure)
+    encoding = tiktoken.encoding_for_model("gpt-4")
+    messages = [SystemMessage(content=context), HumanMessage(content=temp)]
+    if len(encoding.encode(context + temp)) < TOKEN__LIMIT:
```
Comment (kylrth) [inline dataset/format_procedure.py]: Look here for how to count tokens for a chat model: https://github.com/kylrth/procedure-generation/blob/e5638fa0b825f5b94974499f61d63e1346b180b6/systems/model.py#L186

File: dataset/prompt.json (hunk 1 above)
Comment (kylrth) [inline dataset/prompt.json]: format this JSON my guy

File: dataset/format_procedure.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        resp = chatgpt.generate(messages=[messages])
+    else:
+        return ""
```
Comment (kylrth) [inline dataset/format_procedure.py]: We should probably throw an error here, then catch the error in the main function, log it, and skip writing anything in the output file.

File: dataset/format_procedure.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    prompt = json.load(f)
+context = prompt["context"]
+task = prompt["task"]
+TOKEN__LIMIT = 8192
+
+
+def format_procedure(full_procedure: str) -> str:
+    """Formats a single procedure with GPT-4"""
+    temp = task.format(full_procedure)
+    encoding = tiktoken.encoding_for_model("gpt-4")
```
Comment (kylrth) [inline dataset/format_procedure.py]: Let's not initialize the encoder every time the function is called; put this with the other global variables

File: dataset/format_procedure.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    """Formats all files in the directory and saves them to a different dir"""
+    idx = 0
+    for filepath in glob.iglob(path + "/**/*.md", recursive=True):
+        with Path.open(Path(filepath), encoding="utf-8") as file:
+            content = file.read()
+            formatted_proc = format_procedure(content)
+            des = filepath.replace("full", "formatted")
+        if formatted_proc:
+            Path.mkdir(Path(des), parents=True, exist_ok=True)
+            with Path.open(Path(des), "w") as file:
```
Comment (kylrth) [inline dataset/format_procedure.py]: Do we need `encoding="utf-8"` here? I'm not sure how this works but I wanted to ask

File: dataset/format_procedure.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+def count_markdown_files(directory: str) -> int:
+    """Counts markdown files in a directory"""
+    return len(glob.glob(directory + "/**/*.md", recursive=True))
+
+
+def format_docs(path: str) -> None:
+    """Formats all files in the directory and saves them to a different dir"""
+    idx = 0
+    for filepath in glob.iglob(path + "/**/*.md", recursive=True):
```
Comment (kylrth) [inline dataset/format_procedure.py]:
```suggestion
    for idx, filepath in enumerate(glob.iglob(path + "/**/*.md", recursive=True)):
```

**Review by kylrth — CHANGES_REQUESTED** [review]
I would also recommend adding to the dataset/README.md file the instruction to use this script

File: dataset/format_procedure.py (hunk 5 above)
Comment (MkYacine) [reply dataset/format_procedure.py]: I think we do. Some of the procedures caused errors of unknown characters when I was trying to read their content.
Adding this solved that problem

File: dataset/format_procedure.py (hunk 3 above)
Comment (MkYacine) [reply dataset/format_procedure.py]: I see wym, but with the changes I made to the code, the writing in the output file now is done in this method.
So there's no need to throw an error to the main method, I'm just logging the files that exceed the token limit.

File: dataset/prompt.json (hunk 1 above)
Comment (MkYacine) [reply dataset/prompt.json]: will do yo

File: dataset/format_procedure.py (hunk 4 above)
Comment (MkYacine) [reply dataset/format_procedure.py]: Got rid of the encoder all together, now using get_num_tokens()

File: dataset/format_procedure.py [hunk 7]
```diff
@@ -47,24 +52,35 @@ def count_markdown_files(directory: str) -> int:
...
 
 
 if __name__ == "__main__":
-    pass
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-d",
+        "--data-dir",
+        type=str,
+        default="./data",
```
Comment (kylrth) [inline dataset/format_procedure.py]: maybe we can make this `"./docs"` so that it works with where my script puts the files

**Review by kylrth — CHANGES_REQUESTED** [review]

**Review by kylrth — APPROVED** [review]

**Pull request merged**